guides/keras_cv/writing_custom_image_augmentations_with_baseimageaugmentationlayer.py

```diff
@@ -0,0 +1,435 @@
+you may still want to implement your own custom technique.
+KerasCV offers a helpful base class for writing data augmentation layers:
+`BaseImageAugmentationLayer`.
+Any augmentation layer built with `BaseImageAugmentationLayer` will automatically be
+compatible with the KerasCV `RandomAugmentationPipeline` class.
+
+This guide will show you how to implement your own custom augmentation layers using
+`BaseImageAugmentationLayer`.  As an example, we will implement a layer that tints all
+images blue.
+It is recommended that you read (TODO LINK TO OTHER TUTORIAL ONCE POSTED) as a precursor
+to this tutorial.
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+import keras_cv
+from tensorflow.keras import layers
+from keras_cv.utils import preprocessing as preprocessing_utils
+from keras_cv.layers import BaseImageAugmentationLayer
+from keras_cv import bounding_boxes
+import matplotlib.pyplot as plt
+
+
+def imshow(img):
+    img = img.astype(int)
+    plt.axis("off")
+    plt.imshow(img)
+    plt.show()
+
+
+def gallery_show(images):
+    images = images.astype(int)
+    for i in range(9):
+        image = images[i]
+        plt.subplot(3, 3, i + 1)
+        plt.imshow(image.astype("uint8"))
+        plt.axis("off")
+    plt.show()
+
+
+"""## Overview
+
+Image augmentation should operate on a sample-wise basis; not batch-size.
+This is a common mistake many machine learning practicioners make when implementing
+custom techniques.
+`BaseImageAugmentation` offers a set of clean abstractions to make implementing image
+augmentation techniques on a sample wises basis much easier.
+This is done by allowing the end user to override an `augment_image()` method and then
+performing automatic vectorization under the hood.
+
+Most augmentation techniques also must sample from one or more random distributions.
+KerasCV offers an abstraction to make random sampling end user configurable: the
+`FactorSampler` API.
+
+Finally, many augmentation techniques requires some information about the pixel values
+present in the input images.  KerasCV offers the `value_range` API to simplify the handling of this.
+
+In our example, we will use the `FactorSampler` API, the `value_range` API, and
+`BaseImageAugmentationLayer` to implement a robust, configurable, and correct `RandomBlueTint` layer.
+
+## augment_image()
+
+Let's start off with the minimum:
+"""
+
+
+class RandomBlueTint(keras_cv.layers.BaseImageAugmentationLayer):
+    def augment_image(self, image, transformation=None):
+        [*others, blue] = tf.unstack(image, axis=-1)
+        blue = tf.clip_by_value(blue + 100, 0.0, 255.0)
+        return tf.stack([*others, blue], axis=-1)
+
+
+"""Our layer overrides `BaseImageAugmentationLayer.augment_image()`.  This method is
+used to augment images given to the layer.  By default, using `BaseImageAugmentationLayer` gives you a few nice features for free:
+
+- support for both unbatched inputs (WHC Tensor)
+- support for batched inputs (BWHC Tensor)
```

Comment:
Or HWC and BHWC? ([reference](https://keras.io/api/layers/convolution_layers/convolution2d/))